Add APIs to `BlobBuilder` for customizing the underlying byte array et al.

teo-tsirpanis opened the pull request:
Fixes #100418

This PR builds on top of @jaredpar's [branch](https://github.com/jaredpar/runtime/tree/perf-builder) to add APIs for customizing the underlying buffer of a `BlobBuilder`. The chunking logic of `BlobBuilder` was updated to allocate multiple additional chunks with a user-customizable maximum size each. As part of this, we use APIs from `System.Text.Unicode.Utf8` to encode UTF-8 strings, which increases performance and safety, and reduces duplicate code.

teo-tsirpanis committed:
Remove dead code.

teo-tsirpanis committed:
Add extensibility points for custom blob builders.

teo-tsirpanis committed:
Add APIs to `BlobBuilder` for customizing the underlying byte array.

Co-authored-by: Jared Parsons <jared@paranoidcoding.org>

teo-tsirpanis committed:
Add tests.

teo-tsirpanis committed:
Support chunking in `WriteBytes(byte, int)` and when writing streams.

teo-tsirpanis committed:
Use framework methods where available.

teo-tsirpanis committed:
Use framework methods to encode to UTF-8, and support chunking in `WriteUTF8`.

teo-tsirpanis committed:
Zero-initialize buffers returned by `BlobBuilder.ReserveBytes`.

Fixes #99244.

teo-tsirpanis committed:
Remove unnecessary pinning.

teo-tsirpanis committed:
Simplify chunking logic with an `IBufferWriter`-like pattern.

teo-tsirpanis committed:
Fix wrong check order.

And add more tests.

teo-tsirpanis committed:
Use a reasonable default max chunk size.

teo-tsirpanis committed:
Double the size of each new chunk until we get to the max chunk size.

teo-tsirpanis committed:
Check that buffer size is not too small.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/BlobUtilities.cs:
```diff
@@ -114,148 +117,52 @@ public static void WriteGuid(this byte[] buffer, int start, Guid value)
         }
 
-        public static void WriteUTF8(this byte[] buffer, int start, char* charPtr, int charCount, int byteCount, bool allowUnpairedSurrogates)
+        public static void WriteUtf8(ReadOnlySpan<char> source, Span<byte> destination, out int charsRead, out int bytesWritten, bool allowUnpairedSurrogates)
```
We check that `status <= OperationStatus.DestinationTooSmall`, which includes `Completed`.

jkotas commented:
> As part of this, we use APIs from System.Text.Unicode.Utf8 to encode UTF-8 strings, which increases performance and safety, and reduces duplicate code.

Is this needed to introduce the new APIs?

System.Text.Unicode.Utf8 change introduces a new dependency for System.Reflection.Metadata on .NET Framework that will be an extra work to push through the system. It would be better to avoid bundling the two changes together in a single PR.

am11 commented:
> > As part of this, we use APIs from System.Text.Unicode.Utf8 to encode UTF-8 strings, which increases performance and safety, and reduces duplicate code.
> 
> Is this needed to introduce the new APIs?

The other PR is approved and ready to merge https://github.com/dotnet/runtime/pull/111292. After the merge and rebase this branch against main, those commits will disappear.

teo-tsirpanis committed:
Remove dependency to `Microsoft.Bcl.Memory`.

We use the old `WriteUtf8` function for downlevel frameworks, rewritten to have a span-based signature and eliminate unsafe code.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs:
```diff
@@ -253,6 +258,7 @@ public void WriteBytes(byte[] buffer) { }
         public void WriteBytes(byte[] buffer, int start, int byteCount) { }
         public void WriteBytes(System.Collections.Immutable.ImmutableArray<byte> buffer) { }
         public void WriteBytes(System.Collections.Immutable.ImmutableArray<byte> buffer, int start, int byteCount) { }
```
This is not the right place for comments.

teo-tsirpanis committed:
Ensure `BlobBuilder.WriteUTF8`'s loop makes progress.

teo-tsirpanis committed:
Fix tests.

teo-tsirpanis committed:
Clean-up a test.

teo-tsirpanis committed:
Remove `Microsoft.Bcl.Memory` from the solution.

teo-tsirpanis commented:
Switched back to the old `WriteUTF8` implementation for downlevel frameworks, which was rewritten to use spans and remove `unsafe` code. On modern .NET we still use `System.Text.Unicode.Utf8`.

Tests pass locally. This is ready for review.

teo-tsirpanis committed:
Remove unnecessary `unsafe`.

jkotas commented:
> WriteUTF8 implementation for downlevel frameworks, which was rewritten to use spans and remove unsafe code

What''s the performance regression introduced by this rewrite on .NET Framework?

Our primary interest in removing unsafe code is on latest .NET. It is fine to keep unsafe code for .NET Framework if it is required for decent performance.

teo-tsirpanis committed:
Fix misunderstanding of `Capacity` property.

teo-tsirpanis committed:
Update the downlevel `WriteUtf8` to use unsafe code.

teo-tsirpanis commented:
I updated the function to use unsafe code and wrote a benchmark to compare it with my initial safe edition. We cannot compare it with the existing unsafe implementation since the functions don't have the same signature and semantics. The numbers look promising so I switched to the `unsafe` edition:

```

BenchmarkDotNet v0.14.0, Windows 10 (10.0.19045.5796/22H2/2022Update)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.8.1 (4.8.9310.0), X64 RyuJIT VectorSize=256
  DefaultJob : .NET Framework 4.8.1 (4.8.9310.0), X64 RyuJIT VectorSize=256


```
| Method | N   | Mean       | Error    | StdDev   | Ratio | RatioSD |
|------- |---- |-----------:|---------:|---------:|------:|--------:|
| **Safe**   | **16**  |   **276.9 ns** |  **5.38 ns** |  **4.77 ns** |  **1.00** |    **0.02** |
| Unsafe | 16  |   173.1 ns |  3.05 ns |  2.85 ns |  0.63 |    0.01 |
|        |     |            |          |          |       |         |
| **Safe**   | **128** | **1,946.7 ns** | **35.89 ns** | **31.81 ns** |  **1.00** |    **0.02** |
| Unsafe | 128 | 1,210.2 ns | 17.84 ns | 15.81 ns |  0.62 |    0.01 |

<details>
<summary>Benchmark code</summary>

```csharp
// See https://aka.ms/new-console-template for more information
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Runtime.InteropServices;
using System.Text;

BenchmarkRunner.Run<Utf8Bench>();

public class Utf8Bench
{
    public string TestString = null!;

    public byte[] TestBytes = new byte[2048];

    [Params(16, 128)]
    public int N { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        var sb = new StringBuilder();
        for (int i = 0; i < N; i++)
        {
            sb.Append('a');
        }
        for (int i = 0; i < N; i++)
        {
            sb.Append('Θ');
        }
        for (int i = 0; i < N; i++)
        {
            sb.Append("😂");
        }
        TestString = sb.ToString();
        TestBytes = new byte[2048];
    }

    [Benchmark(Baseline = true)]
    public int Safe()
    {
        WriteUtf8Safe(TestString.AsSpan(), TestBytes, out int charsRead, out int bytesWritten, true);
        return charsRead + bytesWritten;
    }

    [Benchmark]
    public int Unsafe()
    {
        WriteUtf8Unsafe(TestString.AsSpan(), TestBytes, out int charsRead, out int bytesWritten, true);
        return charsRead + bytesWritten;
    }

    public static void WriteUtf8Safe(ReadOnlySpan<char> source, Span<byte> destination, out int charsRead, out int bytesWritten, bool allowUnpairedSurrogates)
    {
        // Copy from PR
    }

    public static unsafe void WriteUtf8Unsafe(ReadOnlySpan<char> source, Span<byte> destination, out int charsRead, out int bytesWritten, bool allowUnpairedSurrogates)
    {
        // Copy from PR
    }
}
```
</details>

teo-tsirpanis committed:
Call `OnLinking` earlier.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs:
```diff
@@ -73,6 +182,37 @@ protected virtual void FreeChunk()
+            // nop
+        }
+
+        /// <summary>
+        /// Changes the size of the byte array underpinning the <see cref="BlobBuilder"/>.
+        /// Derived types can override this method to control the allocation strategy.
+        /// </summary>
+        /// <param name="capacity">The array's new size.</param>
+        /// <seealso cref="Capacity"/>
+        protected virtual void SetCapacity(int capacity)
+        {
+            Array.Resize(ref _buffer, Math.Max(MinChunkSize, capacity));
+        }
```
I don't understand how subclasses are supposed to override this. They will reassign the `Buffer` property, but according to a comment in #100418, reassigning `Buffer` clears the head chunk.

What are the semantics? Should `Capacity`'s setter have additional logic?

stephentoub replied:
cc: @jaredpar

teo-tsirpanis committed:
Fix tests on .NET Framework.

We must increment the pointers after we write the bytes.

teo-tsirpanis committed:
Fix reference assembly typos.

teo-tsirpanis committed:
Merge branch 'main' into srm-buffers

stephentoub commented:
> The numbers look promising so I switched to the unsafe edition:

Only for downlevel or everywhere? The numbers you quote are sizeable.

teo-tsirpanis commented:
Only for downlevel. On modern .NET we use `System.Text.Unicode.Utf8`, which looks more optimized. I could vendor the sources to SRM, if we want to use that while avoiding the `Microsoft.Bcl.Memory` dependency.

teo-tsirpanis commented:
I measured the safe `WriteUtf8` implementation I wrote in 515e8259d8382e7b63f5fb34285412f852fe3b2c versus the unsafe one I wrote in 2e8f025267db4738f8ca60210520ce823292e6c2.

teo-tsirpanis committed:
Merge branch 'main' into srm-buffers

AaronRobinsonMSFT commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/BlobUtilities.cs:
```diff
@@ -123,148 +128,135 @@ public static void WriteGuid(this byte[] buffer, int start, Guid value)
+            bytesWritten = destinationLength - destination.Length;
+        }
+#else
+        public static void WriteUtf8(ReadOnlySpan<char> source, Span<byte> destination, out int charsRead, out int bytesWritten, bool allowUnpairedSurrogates)
```
Does this need to change at all? If this is for .NET Framework, I would just leave it as-is. There is little benefit to changing anything due to possibility of regressions.

teo-tsirpanis replied:
Yes. The method's old signature was harder to work with; it uses pointers, does not clearly state which buffer is read-only, and requires pre-computing the UTF-8 byte count.

I had to refactor SRM's UTF-8 encoder[^1]  in order to improve code reusability and memory safety in modern frameworks, and take advantage of the `System.Text.Unicode.Utf8` APIs. Originally, I tried adding a reference to the `Microsoft.Bcl.Memory` package — which has a polyfill to `Utf8`, and this method had a single implementation. Adding the extra dependency to SRM however [is not going to be simple](https://github.com/dotnet/runtime/pull/115294#issuecomment-2850083627), hence the custom code's re-introduction.

Another idea if we want to avoid the extra package dependency, is to vendor [the sources of `Utf8`](https://github.com/dotnet/runtime/blob/b3de022ce057a48a3956c4d691c5b9e4a2714ba7/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj#L30-L44) to SRM. 

[^1]: which has unique requirements, because it is essentially a [WTF-8](https://simonsapin.github.io/wtf-8/) decoder

AaronRobinsonMSFT commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs:
```diff
@@ -109,6 +127,10 @@ public MetadataBuilder(
 
+            _createBlobBuilderFunc = createBlobBuilderFunc ?? (capacity => new BlobBuilder(capacity));
+            _userStringBuilder = _createBlobBuilderFunc(4 * 1024);
+            _guidBuilder = _createBlobBuilderFunc(16); // full metadata has just a single guid
```
```suggestion
            _guidBuilder = _createBlobBuilderFunc(BlobUtilities.SizeOfGuid); // full metadata has just a single guid
```

AaronRobinsonMSFT commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs:
```diff
@@ -109,6 +127,10 @@ public MetadataBuilder(
             }
 
+            _createBlobBuilderFunc = createBlobBuilderFunc ?? (capacity => new BlobBuilder(capacity));
+            _userStringBuilder = _createBlobBuilderFunc(4 * 1024);
```
Is this 4k arbitrary or based on prior art? I'm fine with it, just curious why it was choosen.

AaronRobinsonMSFT reviewed (commented):
Nothing jumps out to me as being the reason for the slowdown. @teo-tsirpanis Have you tried running this under a profiler and seeing where the hot spots are?

jkotas commented:
> We are trying to minimize our use of unsafe throughout, so I expect we'd want to avoid bringing that in here.

The unsafe code is used in .NET Framework polyfills for the most part. There is not much we can do about that without significantly regressing performance of this library on .NET Framework.

These APIs are expected to improve Roslyn performance once Roslyn switches over to use them. Like with any performance related change, I would like to see some numbers that show (1) the refactoring required by these APIs is not regressing Roslyn performance and (2) switching Roslyn to use these APIs is improving Roslyn performance.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/BlobUtilities.cs:
```diff
@@ -123,148 +128,135 @@ public static void WriteGuid(this byte[] buffer, int start, Guid value)
-        }
-
-        internal static unsafe int GetUTF8ByteCount(char* str, int charCount, int byteLimit, out char* remainder)
+#if !NET
```
This is also compiled for .NET Standard 2.0.

teo-tsirpanis committed:
Merge branch 'main' into srm-buffers

jkotas commented:
> It sounds like there's an assumption that we need to bring this new API surface to the .NET Framework polyfill. Why is that the case?

This is a question for Roslyn (@jaredpar). These APIs are designed to optimize Roslyn. If Roslyn is fine with missing these APIs on netfx, we can certainly drop them there.

teo-tsirpanis commented:
> I would like to see the numbers for the implementation in this PR and including walk clock time.

I wrote a benchmark of `BlobBuilder.WriteUTF8` similar to my previous one, comparing between SRM from NuGet and a local build of this branch, running on .NET Framework and .NET 10. I also included writing ASCII-only strings, which is what the overwhelming majority of metadata identifiers are going to be in a compilation.

### .NET Framework

```

BenchmarkDotNet v0.15.8, Windows 10 (10.0.19045.6456/22H2/2022Update)
Intel Core i7-7700HQ CPU 2.80GHz (Max: 2.81GHz) (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.8.1 (4.8.9310.0), X64 RyuJIT VectorSize=256
  DefaultJob : .NET Framework 4.8.1 (4.8.9310.0), X64 RyuJIT VectorSize=256


```
| Method | AsciiOnly | N   | Mean       | Error   | StdDev  |
|------- |---------- |---- |-----------:|--------:|--------:|
| **Write_Before**  | **False**     | **16**  |   **222.5 ns** | **3.81 ns** | **3.38 ns** |
| **Write_Before**  | **True**      | **16**  |   **164.9 ns** | **0.71 ns** | **0.67 ns** |
| **Write_Before**  | **False**     | **128** | **1,474.0 ns** | **7.12 ns** | **6.31 ns** |
| **Write_Before**  | **True**      | **128** |   **993.3 ns** | **5.79 ns** | **5.13 ns** |
| | | | | |
| **Write_After**  | **False**     | **16**  |   **219.7 ns** | **1.62 ns** | **1.26 ns** |
| **Write_After**  | **True**      | **16**  |   **148.4 ns** | **1.09 ns** | **0.91 ns** |
| **Write_After**  | **False**     | **128** | **1,291.5 ns** | **6.24 ns** | **5.53 ns** |
| **Write_After**  | **True**      | **128** |   **698.3 ns** | **4.26 ns** | **3.99 ns** |

### .NET 10

```

BenchmarkDotNet v0.15.8, Windows 10 (10.0.19045.6456/22H2/2022Update)
Intel Core i7-7700HQ CPU 2.80GHz (Max: 2.81GHz) (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 10.0.102
  [Host] : .NET 10.0.2 (10.0.2, 10.0.225.61305), X64 RyuJIT x86-64-v3

Job=InProcess  Toolchain=InProcessEmitToolchain  

```
| Method | AsciiOnly | N   | Mean        | Error    | StdDev   |
|------- |---------- |---- |------------:|---------:|---------:|
| **Write_Before**  | **False**     | **16**  |   **154.05 ns** | **1.515 ns** | **1.265 ns** |
| **Write_Before**  | **True**      | **16**  |    **65.35 ns** | **0.233 ns** | **0.206 ns** |
| **Write_Before**  | **False**     | **128** | **1,159.08 ns** | **7.941 ns** | **7.428 ns** |
| **Write_Before**  | **True**      | **128** |   **467.17 ns** | **2.684 ns** | **2.241 ns** |
| | | | | |
| **Write_After**  | **False**     | **16**  |  **85.51 ns** | **1.687 ns** | **2.675 ns** |
| **Write_After**  | **True**      | **16**  |  **19.53 ns** | **0.146 ns** | **0.114 ns** |
| **Write_After**  | **False**     | **128** | **532.97 ns** | **7.474 ns** | **6.991 ns** |
| **Write_After**  | **True**      | **128** |  **28.57 ns** | **0.308 ns** | **0.257 ns** |

[Sources](https://github.com/user-attachments/files/24987727/srm-bench.zip) are attached. I manually changed the TFM, the `UseStockSrm` property, and the path to the local SRM assembly, and concatenated the results into the tables above.

teo-tsirpanis committed:
Address review comment.

teo-tsirpanis committed:
Address some Copilot comments.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs:
```diff
@@ -61,6 +67,109 @@ public BlobBuilder(int capacity = DefaultChunkSize)
+        /// </summary>
+        /// <remarks>
+        /// This can only be called on the head of a chain of <see cref="BlobBuilder"/> instances.
+        /// Calling the setter will reset the <see cref="Length"/> to zero.
```
I prefer it this way.

teo-tsirpanis committed:
Merge branch 'main' into srm-buffers

teo-tsirpanis committed:
Free the old buffer if `Expand` ends up reusing the same chunk.

teo-tsirpanis committed:
Remove `SetCapacity` and implement setting capacity by resizing chunk in place or expanding it.

This makes the `BlobBuilder` API simpler to extend, at the cost of requiring `OnLinking` to check that blob builders with differing buffer management strategies cannot be mixed.

teo-tsirpanis committed:
Add comments.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs:
```diff
@@ -61,6 +67,109 @@ public BlobBuilder(int capacity = DefaultChunkSize)
+
+                _buffer = value;
+                _length = 0;
+            }
```
Added comment on why this is by design.

teo-tsirpanis committed:
Address Copilot feedback.

teo-tsirpanis committed:
Fix unused usings.

teo-tsirpanis committed:
Merge branch 'main' into srm-buffers

teo-tsirpanis committed:
Do not require the buffer to be at least 16 bytes long in the constructor.

A too small buffer doesn't impact the rest of the class, and BlobBuilder pool implementations using a separate array pool can set the buffer of standby instances to an empty array.

teo-tsirpanis committed:
Update `PooledBlobBuilder` to use buffers from the array pool.

Also fix a gap in the use of `ManagedPEBuilder.CreateBlobBuilder()`.

teo-tsirpanis committed:
Merge branch 'main' into srm-backup

teo-tsirpanis committed:
Fix pooled buffer double freeing in tests.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs:
```diff
@@ -462,24 +462,22 @@ public void WriteUTF8(string value, bool allowUnpairedSurrogates)
-
-                if (prependSize)
-                {
-                    WriteCompressedInteger(byteCount);
-                }
+                WriteCompressedInteger(Encoding.UTF8.GetByteCount(str));
+            }
 
-                int startOffset = Advance(byteCount);
-                _buffer.WriteUTF8(startOffset, charPtr, length, byteCount, allowUnpairedSurrogates);
+            BlobUtilities.WriteUtf8(str.AsSpan(), _buffer.AsSpan(_position), out int charsRead, out int bytesWritten, allowUnpairedSurrogates);
+            if (charsRead != str.Length)
+            {
+                Throw.OutOfBounds();
             }
+            _position += bytesWritten;
         }
```
We already write garbage data, if `prependSize == true` and the buffer is long enough for the size but not the string.

teo-tsirpanis committed:
Address Copilot feedback.

JulieLeeMSFT commented:
@jkotas, PTAL.

jkotas commented:
Converting to draft until we can confirm performance impact on Roslyn.

Also, see the discussion at https://github.com/dotnet/runtime/pull/127304#issuecomment-4320753272

teo-tsirpanis committed:
Merge branch 'main' into srm-buffers
